Sources/Plasma/FeatureLib/pfMetalPipeline/plMetalPipeline.cpp

```diff
@@ -2378,87 +2478,31 @@ void plMetalPipeline::ISelectLights(const plSpan* span, plMetalMaterialShaderRef
     }
 }
 
-void plMetalPipeline::IEnableLight(size_t i, plLightInfo* light)
+void plMetalPipeline::ILoadLight(plLightInfo* light)
 {
-    hsColorRGBA amb = light->GetAmbient();
-    fLights.lampSources[i].ambient = {static_cast<half>(amb.r), static_cast<half>(amb.g), static_cast<half>(amb.b), static_cast<half>(amb.a)};
-
-    hsColorRGBA diff = light->GetDiffuse();
-    fLights.lampSources[i].diffuse = {static_cast<half>(diff.r), static_cast<half>(diff.g), static_cast<half>(diff.b), static_cast<half>(diff.a)};
-
-    hsColorRGBA spec = light->GetSpecular();
-    fLights.lampSources[i].specular = {static_cast<half>(spec.r), static_cast<half>(spec.g), static_cast<half>(spec.b), static_cast<half>(spec.a)};
-
-    plDirectionalLightInfo* dirLight = nullptr;
-    plOmniLightInfo*        omniLight = nullptr;
-    plSpotLightInfo*        spotLight = nullptr;
-    
-    constexpr float         kMaxRange = 32767.f;
-    fLights.lampSources[i].range = kMaxRange;
-
-    if ((dirLight = plDirectionalLightInfo::ConvertNoRef(light)) != nullptr) {
-        hsVector3 lightDir = dirLight->GetWorldDirection();
-        fLights.lampSources[i].position = {lightDir.fX, lightDir.fY, lightDir.fZ, 0.0};
-        fLights.lampSources[i].direction = {lightDir.fX, lightDir.fY, lightDir.fZ};
-
-        fLights.lampSources[i].constAtten = 1.0f;
-        fLights.lampSources[i].linAtten = 0.0f;
-        fLights.lampSources[i].quadAtten = 0.0f;
-
-    } else if ((omniLight = plOmniLightInfo::ConvertNoRef(light)) != nullptr) {
-        hsPoint3 pos = omniLight->GetWorldPosition();
-        fLights.lampSources[i].position = {pos.fX, pos.fY, pos.fZ, 1.0};
-
-        fLights.lampSources[i].constAtten = omniLight->GetConstantAttenuation();
-        fLights.lampSources[i].linAtten = omniLight->GetLinearAttenuation();
-        fLights.lampSources[i].quadAtten = omniLight->GetQuadraticAttenuation();
-
-        if (omniLight->GetRadius() != 0.f) {
-            fLights.lampSources[i].range = omniLight->GetRadius();
-        }
-
-        if (!omniLight->GetProjection() && (spotLight = plSpotLightInfo::ConvertNoRef(omniLight)) != nullptr) {
-            hsVector3 lightDir = spotLight->GetWorldDirection();
-            fLights.lampSources[i].direction = {lightDir.fX, lightDir.fY, lightDir.fZ};
-
-            float falloff = spotLight->GetFalloff();
-            float gamma = cosf(spotLight->GetSpotInner());
-            float phi = cosf(spotLight->GetProjection() ? hsConstants::half_pi<float> : spotLight->GetSpotOuter());
-
-            fLights.lampSources[i].spotProps = {falloff, gamma, phi};
-        } else {
-            fLights.lampSources[i].spotProps = {0.0f, 0.0f, 0.0f};
-        }
-    } else {
-        IDisableLight(i);
-    }
+    auto metalLight = static_cast<plMetalLightRef*>(light->GetDeviceRef());
+    // Track the index of the light for this pass so we can find it again
+    metalLight->fPassIndex = fLights->size();
+    fLights->push_back({metalLight->fBufferIndex,
```

Comment:
```suggestion
    fLights->emplace_back({metalLight->fBufferIndex,
```

Reply:
Hopefully that won't cause a problem. IIRC, according to the standard, this `push_back` call should construct the object, then make a copy of it in the vector, which is not ideal. OTOH `emplace_back` will construct the object in place in the vector. On MSVC, `push_back` just calls `emplace_back`, so you get the most performant path by default, but that's not helpful here. I don't know what other STLs do.